src/platform/datapath_iouring.c

```diff
@@ -0,0 +1,2264 @@
+            RecvData->Allocated = TRUE;
+            RecvData->Route->DatapathType = RecvData->DatapathType = CXPLAT_DATAPATH_TYPE_NORMAL;
+            RecvData->QueuedOnConnection = FALSE;
+            RecvData->Reserved = FALSE;
+
+            *DatagramTail = RecvData;
+            DatagramTail = &RecvData->Next;
+
+            Offset += RecvData->BufferLength;
+            Datagram = (DATAPATH_RX_PACKET*)
+                ((char*)Datagram + SocketContext->DatapathPartition->Datapath->RecvBlockStride);
+        }
+    }
+
+    if (BytesTransferred == 0 || DatagramHead == NULL) {
+        QuicTraceLogWarning(
+            DatapathRecvEmpty,
+            "[data][%p] Dropping datagram with empty payload.",
+            SocketContext->Binding);
+        return;
+    }
+
+    if (CxPlatRundownAcquire(&SocketContext->UpcallRundown)) {
+        if (!SocketContext->Binding->PcpBinding) {
+            CXPLAT_DBG_ASSERT(SocketContext->Binding->Datapath->UdpHandlers.Receive);
+            SocketContext->Binding->Datapath->UdpHandlers.Receive(
+                SocketContext->Binding,
+                SocketContext->Binding->ClientContext,
+                DatagramHead);
+        } else{
+            CxPlatPcpRecvCallback(
+                SocketContext->Binding,
+                SocketContext->Binding->ClientContext,
+                DatagramHead);
+        }
+
+        CxPlatRundownRelease(&SocketContext->UpcallRundown);
+    }
+}
+
+void
+CxPlatSocketReceiveComplete(
+    _In_ CXPLAT_SOCKET_CONTEXT* SocketContext,
+    _In_ CXPLAT_CQE Cqe
+    )
+{
+    CXPLAT_DATAPATH_PARTITION* DatapathPartition = SocketContext->DatapathPartition;
+    DATAPATH_RX_IO_BLOCK* IoBlock;
+    uint8_t* IoPayload;
+    struct msghdr RecvMsgHdrs[1];
+    struct iovec RecvIov;
+    uint32_t BufferIndex;
+    struct io_uring_recvmsg_out* RecvMsgOut;
+
+    if (Cqe->res == -ENOBUFS) {
+        //
+        // Ignore packet loss indications for now.
+        //
+        goto Exit;
+    }
+
+    if (Cqe->res < 0) {
+        if (CxPlatRundownAcquire(&SocketContext->UpcallRundown)) {
+            CxPlatSocketHandleError(SocketContext, -Cqe->res);
+            CxPlatRundownRelease(&SocketContext->UpcallRundown);
+        }
+        goto Exit;
+    }
+
+    CXPLAT_DBG_ASSERT(Cqe->flags & IORING_CQE_F_BUFFER);
+
+    BufferIndex = Cqe->flags >> 16;
+    IoBlock =
+        (DATAPATH_RX_IO_BLOCK*)CxPlatGetBufferPoolBuffer(
+            &DatapathPartition->RecvRegisteredBufferPool, BufferIndex);
+    IoPayload = (uint8_t*)IoBlock + DatapathPartition->Datapath->RecvBlockBufferOffset;
+    RecvMsgOut = io_uring_recvmsg_validate(IoPayload, Cqe->res, (struct msghdr*)&CxPlatRecvMsgHdr);
+    CXPLAT_FRE_ASSERT(RecvMsgOut != NULL); // Review: can this legally fail?
+
+    CXPLAT_DBG_ASSERT((uintptr_t)IoBlock % CXPLAT_MEMORY_ALIGNMENT == 0);
+
+    IoBlock->Route.State = RouteResolved;
+
+    //
+    // Review: these can be batched by propagating the CQE array here.
```

Comment:
We should track this with a backlog task; batching usually provides tangible performance benefits and we should make use of that when possible.

Reply:
IMO let's not clutter up our backlog with these items yet - these are just ideas, hence "review"